File: src/main/java/lambda/evaluator/ProgramStackItemGenerator.java
package lambda.evaluator;

import lambda.core.datamodels.AtomNode;
import lambda.core.datamodels.ExpressionListNode;
import lambda.core.datamodels.Node;
import lambda.core.datamodels.ProgramStackItem;

import java.util.Map;

interface ProgramStackItemGenerator {

    ProgramStackItem generateProgramStackItemFromScratch(
            ExpressionListNode functionExpressionNode,
            Map<String, Node> variableMap
    );

    ProgramStackItem generateProgramStackItemFromExisting(
            ProgramStackItem existingProgramStackItem
    );

    static ProgramStackItemGenerator newInstance() {
        return new ProgramStackItemGenerator() {
            @Override
            public ProgramStackItem generateProgramStackItemFromScratch(ExpressionListNode functionExpressionNode, Map<String, Node> variableMap) {
                return new ProgramStackItem(
                    functionExpressionNode,
                    0,
                    variableMap,
                    ((AtomNode)functionExpressionNode.getChildren().get(0)).getValue()
                );
            }

            @Override
            public ProgramStackItem generateProgramStackItemFromExisting(ProgramStackItem existingProgramStackItem) {
                return new ProgramStackItem(
                    existingProgramStackItem.getFunctionExpressionNode(),
                    existingProgramStackItem.getCurrentParameterIndex() + 1,
                    existingProgramStackItem.getVariableMap(),
                    existingProgramStackItem.getFunctionName()
                );
            }
        };
    }
}

File: src/main/java/lambda/evaluator/ProgramStackItemGenerator.kt
package lambda.evaluator

import lambda.core.datamodels.AtomNode
import lambda.core.datamodels.ExpressionListNode
import lambda.core.datamodels.Node
import lambda.core.datamodels.ProgramStackItem

internal interface ProgramStackItemGenerator {
    fun generateProgramStackItemFromScratch(
        functionExpressionNode: ExpressionListNode,
        variableMap: Map<String, Node>
    ): ProgramStackItem

    fun generateProgramStackItemFromExisting(
        existingProgramStackItem: ProgramStackItem
    ): ProgramStackItem
}

internal class ProgramStackItemGeneratorImpl: ProgramStackItemGenerator {

    override fun generateProgramStackItemFromScratch(
        functionExpressionNode: ExpressionListNode,
        variableMap: Map<String, Node>
    ): ProgramStackItem {
        return ProgramStackItem(
            functionExpressionNode,
            0,
            variableMap,
            (functionExpressionNode.children[0] as AtomNode).value
        )
    }

    override fun generateProgramStackItemFromExisting(
        existingProgramStackItem: ProgramStackItem
    ): ProgramStackItem {
        return ProgramStackItem(
            existingProgramStackItem.functionExpressionNode,
            existingProgramStackItem.currentParameterIndex + 1,
            existingProgramStackItem.variableMap,
            existingProgramStackItem.functionName
        )
    }
}